rcl_action/src/rcl_action/action_server.c

```diff
@@ -0,0 +1,759 @@
+  allocator.deallocate(Type ## _service_name, allocator.state); \
+  if (RCL_RET_OK != ret) { \
+    if (RCL_RET_BAD_ALLOC == ret) { \
+      ret = RCL_RET_BAD_ALLOC; \
+    } else if (RCL_RET_SERVICE_NAME_INVALID == ret) { \
+      ret = RCL_RET_ACTION_NAME_INVALID; \
+    } else { \
+      ret = RCL_RET_ERROR; \
+    } \
+    goto fail; \
+  }
+
+#define PUBLISHER_INIT(Type) \
+  char * Type ## _topic_name = NULL; \
+  ret = rcl_action_get_ ## Type ## _topic_name(action_name, allocator, &Type ## _topic_name); \
+  if (RCL_RET_OK != ret) { \
+    RCL_SET_ERROR_MSG("failed to get " #Type " topic name"); \
+    if (RCL_RET_BAD_ALLOC == ret) { \
+      ret = RCL_RET_BAD_ALLOC; \
+    } else { \
+      ret = RCL_RET_ERROR; \
+    } \
+    goto fail; \
+  } \
+  rcl_publisher_options_t Type ## _publisher_options = { \
+    .qos = options->Type ## _topic_qos, .allocator = allocator \
+  }; \
+  action_server->impl->Type ## _publisher = rcl_get_zero_initialized_publisher(); \
+  ret = rcl_publisher_init( \
+    &action_server->impl->Type ## _publisher, \
+    node, \
+    type_support->Type ## _message_type_support, \
+    Type ## _topic_name, \
+    &Type ## _publisher_options); \
+  allocator.deallocate(Type ## _topic_name, allocator.state); \
+  if (RCL_RET_OK != ret) { \
+    if (RCL_RET_BAD_ALLOC == ret) { \
+      ret = RCL_RET_BAD_ALLOC; \
+    } else if (RCL_RET_TOPIC_NAME_INVALID == ret) { \
+      ret = RCL_RET_ACTION_NAME_INVALID; \
+    } else { \
+      ret = RCL_RET_ERROR; \
+    } \
+    goto fail; \
+  }
+
+rcl_ret_t
+rcl_action_server_init(
+  rcl_action_server_t * action_server,
+  rcl_node_t * node,
+  const rosidl_action_type_support_t * type_support,
+  const char * action_name,
+  const rcl_action_server_options_t * options)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(action_server, RCL_RET_INVALID_ARGUMENT);
+  if (!rcl_node_is_valid(node)) {
+    return RCL_RET_NODE_INVALID;  // error already set
+  }
+  RCL_CHECK_ARGUMENT_FOR_NULL(type_support, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(action_name, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(options, RCL_RET_INVALID_ARGUMENT);
+  rcl_allocator_t allocator = options->allocator;
+  RCL_CHECK_ALLOCATOR_WITH_MSG(&allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+
+  RCUTILS_LOG_DEBUG_NAMED(
+    ROS_PACKAGE_NAME, "Initializing action server for action name '%s'", action_name);
+  if (action_server->impl) {
+    RCL_SET_ERROR_MSG("action server already initialized, or memory was unintialized");
+    return RCL_RET_ALREADY_INIT;
+  }
+
+  // Allocate for action server
+  action_server->impl = (rcl_action_server_impl_t *)allocator.allocate(
+    sizeof(rcl_action_server_impl_t), allocator.state);
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    action_server->impl, "allocating memory failed", return RCL_RET_BAD_ALLOC);
+
+  // Initialize clock
+  rcl_ret_t ret = rcl_clock_init(options->clock_type, &action_server->impl->clock, &allocator);
```

Comment:
Okay. Then I suppose it will not be the action servers responsibility to finalize the clock and we hope that it stays valid as long as the action server is valid. Does this sound right?

Reply:
> Does this sound right?

I think so

Reply:
Addressed in 4eb7fdb